libRALFit/src/ral_nlls_internal.f90

```diff
@@ -3928,27 +3949,30 @@ Recursive subroutine solve_newton_tensor(J, f, eval_HF, X, n, m, Delta, &
          goto 100
        End If
 
-       ! save to params
-       w%tparams%f(1:m) = f(1:m)
-       w%tparams%Delta = Delta
-       w%tparams%J(1:n*m) = J(1:n*m)
-       w%tparams%X(1:n) = X(1:n)
-       w%tparams%eval_HF => eval_HF
-       w%tparams%parent_params => params
-       w%tparams%tenJ => tenJ
-
        md = 0.0_wp
 
        if (.not. w%tparams%eval_hp_provided) then
           ! let's get all the Hi's...
-          do i = 1,m
-             call get_Hi(n, m, X, params, i, w%tparams%Hi(:,:,i), eval_HF, inform)
+          w%tparams%f(:) = 0.0_wp
+          do i = 1, m
+             ! get_Hi expects ei=w%tparams%f to be a zero-filled vector
+             call get_Hi(n, m, X, params, i, w%tparams%f, w%tparams%Hi(:,:,i), &
+               eval_HF, inform)
              If (inform%status/=0) Then
                Go To 100
              End If
           end do
        end if
 
+       ! save to params
+       w%tparams%f(1:m) = f(1:m)
+       w%tparams%Delta = Delta
+       w%tparams%J(1:n*m) = J(1:n*m)
+       w%tparams%X(1:n) = X(1:n)
+       w%tparams%eval_HF => eval_HF
+       w%tparams%parent_params => params
+       w%tparams%tenJ => tenJ
+
```

Comment:
`get_Hi` now has an extra array `ei(m)`, a work array, so it does not require the automatic array equivalent. The idea is to use `w%tparams%f(1:m)` just before it is initialized with `f(1:m)`. In order to exploit this, the initialization must be postponed after the call to `get_Hi` (loop).